zebra-state/src/service/check.rs

```diff
@@ -27,30 +40,51 @@ pub(crate) fn block_is_contextually_valid<C>(
 ) -> Result<(), ValidateContextError>
 where
     C: IntoIterator,
-    C::Item: AsRef<Block>,
+    C::Item: Borrow<Block>,
+    C::IntoIter: ExactSizeIterator,
 {
     let finalized_tip_height = finalized_tip_height
-        .expect("finalized state must contain at least one block to use the non-finalized state");
+        .expect("finalized state must contain at least one block to do contextual validation");
     check::block_is_not_orphaned(finalized_tip_height, prepared.height)?;
 
-    let mut relevant_chain = relevant_chain.into_iter();
+    // The maximum number of blocks used by contextual checks
+    const MAX_CONTEXT_BLOCKS: usize = POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN;
+    let relevant_chain: Vec<_> = relevant_chain
+        .into_iter()
+        .take(MAX_CONTEXT_BLOCKS)
+        .collect();
+
     let parent_block = relevant_chain
-        .next()
+        .get(0)
         .expect("state must contain parent block to do contextual validation");
-    let parent_block = parent_block.as_ref();
+    let parent_block = parent_block.borrow();
     let parent_height = parent_block
         .coinbase_height()
         .expect("valid blocks have a coinbase height");
     check::height_one_more_than_parent_height(parent_height, prepared.height)?;
 
-    // TODO: validate difficulty adjustment
-    // TODO: other contextual validation design and implelentation
+    if relevant_chain.len() >= POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN {
```

Comment:
Overall I'm not thrilled with having this change here. It's here to support https://github.com/ZcashFoundation/zebra/pull/1354, which is a change we haven't yet decided to merge, so if we end up not merging that other PR this is going to be a liability. My preference would be to remove this change from here and move it to #1354, so it can be part of the discussions that are relevant to it.

Reply:
Fair enough, I put the commit in here to avoid a merge conflict with #1354, because I didn't know which order they'd merge in. (And GitHub squashes merges by default, which makes it much harder to automatically resolve conflicts.)

I've turned this check into an assert in 0466b2f25acc0ca6607ce805dcc03f21272f07b3, and I'll base #1354 on this PR.